validation/validation_test.go

```diff
@@ -24,81 +27,106 @@ func init() {
 	}
 }
 
-func runtimeValidate(runtime string, g *generate.Generator) error {
-	// Find the runtime binary in the PATH
-	runtimePath, err := exec.LookPath(runtime)
+func prepareBundle() (string, error) {
+	// Setup a temporary test directory
+	bundleDir, err := ioutil.TempDir("", "ocitest")
 	if err != nil {
-		return err
+		return "", err
 	}
 
-	// Setup a temporary test directory
-	tmpDir, err := ioutil.TempDir("", "ocitest")
+	// Untar the root fs
+	untarCmd := exec.Command("tar", "-xf", "../rootfs.tar.gz", "-C", bundleDir)
+	_, err = untarCmd.CombinedOutput()
 	if err != nil {
-		return err
+		os.RemoveAll(bundleDir)
+		return "", err
 	}
-	defer os.RemoveAll(tmpDir)
 
-	// Create bundle directory for the test container
-	bundleDir := tmpDir + "/busybox"
-	if err := os.MkdirAll(bundleDir, 0755); err != nil {
+	return bundleDir, nil
+}
+
+func getDefaultGenerator() *generate.Generator {
+	g := generate.New()
+	g.SetRootPath(".")
+	g.SetProcessArgs([]string{"/runtimetest"})
+	return &g
+}
+
+func runtimeInsideValidate(g *generate.Generator) error {
+	bundleDir, err := prepareBundle()
+	if err != nil {
 		return err
 	}
-
-	// Untar the root fs
-	untarCmd := exec.Command("tar", "-xf", "../rootfs.tar.gz", "-C", bundleDir)
-	output, err := untarCmd.CombinedOutput()
+	r, err := NewRuntime(runtime, bundleDir)
 	if err != nil {
-		fmt.Println(string(output))
+		os.RemoveAll(bundleDir)
 		return err
 	}
-
-	// Copy the runtimetest binary to the rootfs
-	err = fileutils.CopyFile("../runtimetest", filepath.Join(bundleDir, "runtimetest"))
+	defer r.Clean(true)
+	err = r.SetConfig(g)
 	if err != nil {
 		return err
 	}
-
-	// Generate test configuration
-	err = g.SaveToFile(filepath.Join(bundleDir, "config.json"), generate.ExportOptions{})
+	err = fileutils.CopyFile("../runtimetest", filepath.Join(r.BundleDir, "runtimetest"))
 	if err != nil {
 		return err
 	}
 
-	// TODO: Use a library to split run into create/start
-	// Launch the OCI runtime
-	containerID := uuid.NewV4()
-	runtimeCmd := exec.Command(runtimePath, "run", containerID.String())
-	runtimeCmd.Dir = bundleDir
-	runtimeCmd.Stdin = os.Stdin
-	runtimeCmd.Stdout = os.Stdout
-	runtimeCmd.Stderr = os.Stderr
-	if err = runtimeCmd.Run(); err != nil {
+	r.SetID(uuid.NewV4().String())
+	err = r.Create()
+	if err != nil {
 		return err
 	}
-
-	return nil
-}
-
-func getDefaultGenerator() *generate.Generator {
-	g := generate.New()
-	g.SetRootPath(".")
-	g.SetProcessArgs([]string{"/runtimetest"})
-	return &g
+	return r.Start()
```

Comment:
This is broken, since you've deferred a `delete` call (via `Clean`) [earlier in the function][1].  That doesn't work reliably though, because [`delete` MUST generate an error if the container is not `stopped`][2], and the fact that `start` completes doesn't mean you're `stopped` (you [could also be `running`][3]).  In order for `Clean()` to work reliably, you either need to attempt to send a `KILL` to the container process with [`kill`][4] (although the spec is not clear on what signals are supported, you need something like #321 to see that [`KILL` MUST be supported][5]) or code that waits for the container process to exit on its own (more on this [here][6], in #305, and in most of the other PRs that attempted to split `run` into `create` and`start`, e.g. [here][7] and [here][8]).

[1]: https://github.com/opencontainers/runtime-tools/pull/447/files#diff-b5e8ff579203d652582cf33ed2b39ad9R65
[2]: https://github.com/opencontainers/runtime-spec/blob/v1.0.0/runtime.md#delete
[3]: https://github.com/opencontainers/runtime-spec/blame/v1.0.0/runtime.md#L21-L22
[4]: https://github.com/opencontainers/runtime-spec/blob/v1.0.0/runtime.md#kill
[5]: https://github.com/wking/ocitools-v2/blob/a2725b42d7b312681ed5315cf7e663a13c487445/docs/command-line-interface.md#kill
[6]: https://github.com/wking/ocitools-v2/blob/a2725b42d7b312681ed5315cf7e663a13c487445/docs/command-line-interface.md#container-process-exit
[7]: https://github.com/opencontainers/runtime-tools/pull/232#discussion_r82655833
[8]: https://github.com/opencontainers/runtime-tools/pull/285#issuecomment-264515242